packages/remote/src/electron-node/setup/remote-node-setup-service.ts

```diff
@@ -0,0 +1,89 @@
+//
+// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0
+// *****************************************************************************
+
+import * as path from 'path';
+import * as fs from '@theia/core/shared/fs-extra';
+import * as os from 'os';
+
+import { inject, injectable } from '@theia/core/shared/inversify';
+import { RemotePlatform } from '@theia/core/lib/node/remote';
+import { RequestService } from '@theia/core/shared/@theia/request';
+import { RemoteSetupScriptService } from './remote-setup-script-service';
+
+/**
+ * The current node version that Theia recommends.
+ *
+ * Native dependencies are compiled against this version.
+ */
+export const REMOTE_NODE_VERSION = '18.17.0';
+
+@injectable()
+export class RemoteNodeSetupService {
+
+    @inject(RequestService)
+    protected readonly requestService: RequestService;
+
+    @inject(RemoteSetupScriptService)
+    protected readonly scriptService: RemoteSetupScriptService;
+
+    getNodeDirectoryName(platform: RemotePlatform): string {
+        const platformId = platform === 'windows' ? 'win' : platform;
+        // Always use x64 architecture for now
+        const arch = 'x64';
+        const dirName = `node-v${REMOTE_NODE_VERSION}-${platformId}-${arch}`;
+        return dirName;
+    }
+
+    getNodeFileName(platform: RemotePlatform): string {
+        let fileExtension = '';
+        if (platform === 'windows') {
+            fileExtension = 'zip';
+        } else if (platform === 'darwin') {
+            fileExtension = 'tar.gz';
+        } else {
+            fileExtension = 'tar.xz';
+        }
+        return `${this.getNodeDirectoryName(platform)}.${fileExtension}`;
+    }
+
+    async downloadNode(platform: RemotePlatform): Promise<string> {
+        const fileName = this.getNodeFileName(platform);
+        const tmpdir = os.tmpdir();
+        const localPath = path.join(tmpdir, fileName);
+        if (!await fs.pathExists(localPath)) {
+            const downloadPath = this.getDownloadPath(fileName);
+            const downloadResult = await this.requestService.request({
+                url: downloadPath
+            });
+            await fs.writeFile(localPath, downloadResult.buffer);
+        }
+        return localPath;
+    }
+
+    generateDownloadScript(platform: RemotePlatform, targetPath: string): string {
+        const fileName = this.getNodeFileName(platform);
+        const downloadPath = this.getDownloadPath(fileName);
+        const zipPath = RemotePlatform.joinPath(platform, targetPath, fileName);
+        const download = this.scriptService.downloadFile(platform, downloadPath, zipPath);
+        const unzip = this.scriptService.unzip(zipPath, targetPath);
+        return this.scriptService.joinScript(platform, download, unzip);
+    }
+
+    protected getDownloadPath(fileName: string): string {
+        return `https://nodejs.org/dist/v${REMOTE_NODE_VERSION}/${fileName}`;
```

Comment:
I think how we get this uri should be configurable: think of companies like Ericsson that have environments without access to the public internet. Also for performance and uptime reasons, they might prefer a local copy.

Reply:
For now adopters can just override the variable - making it entirely user configurable is a bit difficult, as the exact download location depends on the remote platform and architecture.

Reply:
Having a hard-coded URL just makes me cringe. If the URL ever changes, we'll have to fix old versions of Theia. Wouldn't a simple template string we read from configuration (preferences?) take care of most use cases?

Reply:
Alright, the URL is now configurable 👍